Auto-generate help message based on inspected types

ahrens opened the pull request:
We can make it easier to discover the features of various commands by
having the "help" message include information that we can discover
programmatically by inspecting the Command subclass.  Specifically:

1. Its input_type
2. Whether it's a PrettyPrinter
3. Whether it's a Walker
4. If it's a Locator, and what types it can locate based on

e.g.:

```
sdb> help spa
SUMMARY
    spa [-h] [-v] [-m] [-H] [-w] [poolnames [poolnames ...]]

    positional arguments:
      poolnames

    optional arguments:
      -h, --help       show this help message and exit
      -v, --vdevs      vdevs flag
      -m, --metaslab   metaslab flag
      -H, --histogram  histogram flag
      -w, --weight     weight flag

    If this command is used to end a pipeline, it will print a human-
    readable decoding of the 'spa_t *' objects. For the 'raw' object
    contents, pipe the output of this command into 'echo'.

    This command accepts inputs of type 'void *', and 'spa_t', which
    will be converted to 'spa_t *'.

    This is a Locator for spa_t *.  It finds objects of this type and
    outputs or pretty-prints them.  It accepts any Walkable type (run
    'walk' for a list). This command can be used to start a pipeline,
    in which case it will consume no objects as input; instead it will
    locate all objects of type 'spa_t *', and emit them as output.

sdb> help avl
SUMMARY
    avl [-h]

    walk avl tree

    optional arguments:
      -h, --help  show this help message and exit

    This is a Walker for avl_tree_t *. See 'help walk'.

    This command accepts inputs of type 'void *', and 'avl_tree_t',
    which will be converted to 'avl_tree_t *'.


sdb> help dbuf
SUMMARY
    dbuf [-h] [-o OBJECT] [-l LEVEL] [-b BLKID] [-d DATASET] [-H]

    Iterate, filter, and pretty-print dbufs (dmu_buf_impl_t*)

    optional arguments:
      -h, --help            show this help message and exit
      -o OBJECT, --object OBJECT
                            filter: only dbufs of this object
      -l LEVEL, --level LEVEL
                            filter: only dbufs of this level
      -b BLKID, --blkid BLKID
                            filter: only dbufs of this blkid
      -d DATASET, --dataset DATASET
                            filter: only dbufs of this dataset name (or
                            "poolname/_MOS")
      -H, --has-holds       filter: only dbufs that have nonzero holds

    If this command is used to end a pipeline, it will print a human-
    readable decoding of the 'dmu_buf_impl_t *' objects. For the 'raw'
    object contents, pipe the output of this command into 'echo'.

    This command accepts inputs of type 'void *', and
    'dmu_buf_impl_t', which will be converted to 'dmu_buf_impl_t *'.

    This is a Locator for dmu_buf_impl_t *.  It finds objects of this
    type and outputs or pretty-prints them.  It accepts any Walkable
    type (run 'walk' for a list). This command can be used to start a
    pipeline, in which case it will consume no objects as input;
    instead it will locate all objects of type 'dmu_buf_impl_t *', and
    emit them as output. Input of the following types is also
    accepted, in which case the objects of type dmu_buf_impl_t * which
    are associated with them will be located:
        dnode_t*


sdb> help vdev
SUMMARY
    vdev [-h] [-m] [-H] [-w] [vdev_ids [vdev_ids ...]]

    positional arguments:
      vdev_ids

    optional arguments:
      -h, --help       show this help message and exit
      -m, --metaslab   metaslab flag
      -H, --histogram  histogram flag
      -w, --weight     weight flag

    If this command is used to end a pipeline, it will print a human-
    readable decoding of the 'vdev_t *' objects. For the 'raw' object
    contents, pipe the output of this command into 'echo'.

    This command accepts inputs of type 'void *', and 'vdev_t', which
    will be converted to 'vdev_t *'.

    This is a Locator for vdev_t *.  It finds objects of this type and
    outputs or pretty-prints them.  It accepts any Walkable type (run
    'walk' for a list). Input of the following types is also accepted,
    in which case the objects of type vdev_t * which are associated
    with them will be located:
        spa_t*
        vdev_t*

sdb> help pp
SUMMARY
    pp [-h]

    optional arguments:
      -h, --help  show this help message and exit

ALIASES
    pretty_print, pp

sdb> help walk
SUMMARY
    walk [-h]

    Dispatch the appropriate walker based on the type of input

    optional arguments:
      -h, --help  show this help message and exit

DESCRIPTION
    This command can be used to walk data structures when
    a specific walker for them already exists. There are
    two scenarios when this command is preferable to using
    a specific walker:
    [1] When objects of different types are passed at once
        through a pipe, this walker can dispatch the
        appropriate walker for each of them so the user
        won't need to care about the underlying data
        structure implementations.
    [2] Commands that depend on a data structure being
        traversed can use this command, to reduce the
        lines of code changed when the underlying data
        structure chages.

    For a list of walkers, run 'walk' with no input.

EXAMPLES

    sdb> addr spa_namespace_avl | walk
    (void *)0xffff9d0adbe2c000
    (void *)0xffff9d0a2dd28000
    (void *)0xffff9d0ae5040000
    (void *)0xffff9d0a2bdb0000

sdb> help range_seg
SUMMARY
    range_seg [-h]

    Locate the range seg's associated with a range tree.

    optional arguments:
      -h, --help  show this help message and exit

    This is a Locator for None.  It finds objects of this type and
    outputs or pretty-prints them.  It accepts any Walkable type (run
    'walk' for a list). Input of the following types is also accepted,
    in which case the objects of type None which are associated with
    them will be located:
        range_tree_t *

Given a 'range_tree_t*', locate the range_seg's assocated with it.
These may be type range_seg32_t*, range_seg64_t*, or range_seg_gap_t*,
depending on what kind of range_tree_t this is.

sdb>
```

ahrens commented:
I think some of the wording here could be improved.  Open to suggestions...

prakashsurya commented on sdb/command.py:
```diff
@@ -73,6 +74,50 @@ def help(cls, name: str):
+            print()
+
+        if issubclass(cls, sdb.Walker):
+            print("PRETTY PRINTER")
```
typo.. should be `WALKER`, right?

prakashsurya commented on sdb/command.py:
```diff
@@ -73,6 +74,50 @@ def help(cls, name: str):
+        if issubclass(cls, sdb.Walker):
+            print("PRETTY PRINTER")
+            print(f"    This is a Walker for {cls.input_type}.  "
+                  "See 'help walk'.")
```
What information do we expect `help walk` to convey to the user for walkers? e.g. if I use `help avl`, and then `help walk`.. `help walk` doesn't really provide any more information to the user..

Do you think the lack of additional information is just because the docstring for `walk` is non-existent?

ahrens replied:
I was thinking we should add the list of walkers, as well as a general description, to `help walk`.

prakashsurya commented on sdb/command.py:
```diff
@@ -73,6 +74,50 @@ def help(cls, name: str):
+        if issubclass(cls, sdb.PrettyPrinter):
+            print("PRETTY PRINTER")
+            print(f"    This is a PrettyPrinter for {cls.input_type}.")
+            print(f"    If prints a human-readable decoding of the object.")
```
"If" -> "It"

prakashsurya commented:
I'm still not done, but what do you think about what I have so far.. e.g.
```
sdb> help spa
SUMMARY
    spa [-h] [-v] [-m] [-H] [-w] [poolnames [poolnames ...]]

    positional arguments:
      poolnames

    optional arguments:
      -h, --help       show this help message and exit
      -v, --vdevs      vdevs flag
      -m, --metaslab   metaslab flag
      -H, --histogram  histogram flag
      -w, --weight     weight flag

    This command can be used to start a pipeline, in which case it
    will consume no objects as input; instead it will locate all
    objects of type 'spa_t *', and emit them as output.

    This command can also be used within a pipeline, such that it
    accepts objects of type 'spa_t *' as input, and will emit objects
    of type 'spa_t *' as output.

    Lastly, if this command is used to end a pipeline, it will print a
    human-readable decoding of the 'spa_t *' objects. For the 'raw'
    object contents, pipe the output of this command into 'echo'.
```
?

Implemented via something like this:
```
 73         summary = []                                                            
 74         if issubclass(cls, sdb.Locator):                                        
 75             summary.append(                                                     
 76                 f"This command can be used to start a pipeline, in which"       
 77                 f" case it will consume no objects as input; instead it"        
 78                 f" will locate all objects of type '{cls.output_type}',"        
 79                 f" and emit them as output."                                    
 80             )                                                                   
 81                                                                                 
 82         if cls.input_type is not None:                                          
 83             summary.append(                                                     
 84                 f"This command can also be used within a pipeline,"             
 85                 f" such that it accepts objects of type '{cls.input_type}'"     
 86                 f" as input, and will emit objects of type '{cls.output_type}'" 
 87                 f" as output."                                                  
 88             )                                                                   
 89                                                                                 
 90         if issubclass(cls, sdb.PrettyPrinter):                                  
 91             summary.append(                                                     
 92                 f"Lastly, if this command is used to end a pipeline, it will"   
 93                 f" print a human-readable decoding of the '{cls.input_type}'"   
 94                 f" objects. For the 'raw' object contents, pipe the output"     
 95                 f" of this command into 'echo'."                                
 96             )                                                                   
 97                                                                                 
 98         indent = "    "                                                         
 99         wrapper = textwrap.TextWrapper(initial_indent=indent,                   
100                                        subsequent_indent=indent)                
101         for paragraph in summary:                                               
102             for line in wrapper.wrap(paragraph):                                
103                 print(line)                                                     
104             print() 
```

This is still missing some of the information that you're currently printing in your `sdb.Walker` and `sdb.Locator` sections..

Do you think it's better to maintain the `PRETTY PRINTER`, `LOCATOR`, etc. section headers like you're currently doing? So far, I think those sections might not be very intuitive to a new user, but might be better suited for an experienced user..

ahrens commented:
@prakashsurya I've updated the code to be similar to what you suggested, and updated the examples in the first comment.  Also made a few minor improvements to other commands help messages.  `help range_seg` is still a bit strange (with `None`) but it's such an edge case that I'm not sure it's worth fixing in the common code.
